ci/pipeline/ray_release_test.py

```diff
@@ -21,19 +23,68 @@ def main() -> int:
     changed_files = _get_changed_files()
     logger.info(f"Changed files: {changed_files}")
     test_targets = _get_test_targets_for_changed_files(changed_files)
-    logger.info(test_targets)
+    logger.info(f"Test targets to run: {test_targets}")
+    if not test_targets:
+        logger.info("No tests to run")
+        return 0
+    _run_tests(test_targets)
```

Comment:
I just really worry that this won't run the right set of tests.. and we are immaturely optimizing things..

could you maybe write some unit tests, which can also serve as examples of how things would work? like:

- what happens when a new test file is added?
- what happens when a test file is renamed?
- what happens when a test file is deleted?
- what happens when a new source file is added?
- what happens when a source file is renamed?
- what happens when a source file is deleted?

Reply:
Absolutely, this deserve some unit-tests. You spotted on some of the cases where this will fall short (e.g. when a new test is added), which I need to handle separately, maybe in a follow up PRs ;)

Reply:
I added unit-tests but not sure if it really answers your questions, so answer it separately here:

- what happens when a test file is renamed?
- what happens when a test file is deleted?
- what happens when a source file is renamed?
- what happens when a source file is deleted?

> these cases are treated the same; since the original test file and source will be in the changed list, the right tests will be run as usual


- what happens when a new source file is added?

> the new source file, without being used anywhere, should not run any tests; if it used in another file, the change of that file will trigger the right tests

- what happens when a new test file is added?

> we should be able to detect that a new file is added and run it accordingly


In theory, a drawback of dynamic coverage is that the coverage information might get staled so tests might be skip incorrectly if new changes change the coverage. In practice, this could be minimal due to the frequency of collecting coverage and the infrequency of test breakages.